modules/SofaMiscFem/src/SofaMiscFem/TriangleFEMForceField.inl

```diff
@@ -497,117 +371,102 @@ void TriangleFEMForceField<DataTypes>::initLarge()
 
-    // Rotation matrix (deformed and displaced Triangle/world)
-    Transformation R_2_0, R_0_2;
-    computeRotationLarge( R_0_2, p, a, b, c);
+        // positions of the deformed points in the local frame
+        const Coord deforme_b = R_0_2 * (pB - pA);
+        const Coord deforme_c = R_0_2 * (pC - pA);
 
+        // displacements in the local frame
+        Displacement Depl;
+        m_triangleUtils->computeDisplacementLarge(Depl, R_0_2, _rotatedInitialElements[elementIndex], pA, pB, pC);
 
-    // positions of the deformed points in the local frame
-    Coord deforme_a, deforme_b, deforme_c;
-    deforme_b = R_0_2 * (p[b]-p[a]);
-    deforme_c = R_0_2 * (p[c]-p[a]);
+        // Strain-displacement matrix
+        StrainDisplacement J;
+        m_triangleUtils->computeStrainDisplacementLocal(J, deforme_b, deforme_c);
 
-    // displacements in the local frame
-    Displacement D;
-    D[0] = 0;
-    D[1] = 0;
-    D[2] = _rotatedInitialElements[elementIndex][1][0] - deforme_b[0];
-    D[3] = 0;
-    D[4] = _rotatedInitialElements[elementIndex][2][0] - deforme_c[0];
-    D[5] = _rotatedInitialElements[elementIndex][2][1] - deforme_c[1];
+        // compute strain
+        type::Vec<3, Real> strain;
+        m_triangleUtils->computeStrain(strain, J, Depl, false);
 
-    // Strain-displacement matrix
-    StrainDisplacement B;
-    computeStrainDisplacement(B,deforme_a,deforme_b,deforme_c);
+        // compute stress
+        type::Vec<3, Real> stress;
+        m_triangleUtils->computeStress(stress, _materialsStiffnesses[elementIndex], strain, false);
 
-    // compute force on element, in local frame
-    Displacement F;
-    computeForce( F, D, _materialsStiffnesses[elementIndex], B ); // F = Bt.S.B.D
+        // compute force on element, in local frame
+        Displacement F;
+        m_triangleUtils->computeForceLarge(F, J, stress);
 
-    // project forces to world frame
-    R_2_0.transpose(R_0_2);
-    f[a] += R_2_0 * Coord(F[0], F[1], 0);
-    f[b] += R_2_0 * Coord(F[2], F[3], 0);
-    f[c] += R_2_0 * Coord(F[4], F[5], 0);
+        // project forces to world frame
+        R_2_0.transpose(R_0_2);
+        f[a] += R_2_0 * Coord(F[0], F[1], 0);
+        f[b] += R_2_0 * Coord(F[2], F[3], 0);
+        f[c] += R_2_0 * Coord(F[4], F[5], 0);
 
-    // store for re-use in matrix-vector products
-    if(implicit)
-    {
-        _strainDisplacements[elementIndex] = B;
-        _rotations[elementIndex] = R_2_0 ;
+        // store for re-use in matrix-vector products
+        if (implicit)
+        {
+            _strainDisplacements[elementIndex] = J;
+            _rotations[elementIndex] = R_2_0;
+        }
     }
-
 }
 
 
 template <class DataTypes>
-void TriangleFEMForceField<DataTypes>::applyStiffnessLarge(VecCoord &v, Real h, const VecCoord &x, const SReal &kFactor)
+void TriangleFEMForceField<DataTypes>::applyStiffnessLarge(VecCoord& v, Real h, const VecCoord& x, const SReal& kFactor)
```

Comment:
This function (and its equivalent) must take `kFactor` as a `Real` and not `SReal`.
Also, for primitive types, passing by const-ref is counterproductive. So pass it by value.

Reply:
you are right thanks!